### ISSUE TITLE
AXI exclusive write fails after regular write from reserving manager

### ISSUE THREAD

**Issue by maxbjurling** [pulp-platform/axi_riscv_atomics#29]
A regular write to a reserved address from the manager that reserved it causes the reservation to be cleared and a subsequent exclusive write to fail. This is not according to the AXI or RISC-V specifications, where only a write from *another* manager should clear the reservation (some other conditions in the RISC-V spec may also cause an exclusive write to fail).

*Note: this is a copy of [Issue 397](https://github.com/pulp-platform/pulpissimo/issues/397) submitted to the pulpissimo repo.*

### Specifications

The [AMBA AXI spec](https://developer.arm.com/documentation/ihi0022/j?lang=en), section _A7.3.1 Exclusive access sequence_, states:

> 3. This exclusive write access is signaled as either:
> - Successful, if no other Manager has written to that location since the exclusive read access. In this case, the exclusive write updates memory.
> - Failed, if another Manager has written to that location since the exclusive read access. In this case, the memory location is not updated.

The [RISC-V spec](https://github.com/riscv/riscv-isa-manual/releases/download/Ratified-IMAFDQC/riscv-spec-20191213.pdf), section *8.2 Load-Reserved/Store-Conditional Instructions*, states (bullets added for readability):

> - An SC can only pair with the most recent LR in program order.
> - An SC may succeed only if no store from another hart to the reservation set can be observed to have occurred between the LR and the SC, and if there is no other SC between the LR and itself in program order.
> - An SC may succeed only if no write from a device other than a hart to the bytes accessed by the LR instruction can be observed to have occurred between the LR and SC. Note this LR might have had a different effective address and data size, but reserved the SC’s address as part of the reservation set.
> - The SC must fail if the address is not within the reservation set of the most recent LR in program order.
> - The SC must fail if a store to the reservation set from another hart can be observed to occur between the LR and SC.
> - The SC must fail if a write from some other device to the bytes accessed by the LR can be observed to occur between the LR and SC. (If such a device writes the reservation set but does not write the bytes accessed by the LR, the SC may or may not fail.)
> - An SC must fail if there is another SC (to any address) between the LR and the SC in program order.

### Reproducing the bug
To trigger this bug, submit the following sequence to the axi_riscv_atomics module:

1. Send an exclusive read with ARID *I* to address *A*
2. Observe RRESP, should be `EXOKAY` to signal successful reservation
3. Send a regular write with AWID *I* to address *A*
4. Send an exclusive write with AWID *I* to address *A*
5. Observe BRESP, should be `EXOKAY` to signal succesful store (but it isn't!)

**Comment by SamuelRiedel** [comment]
You are right, and we can look into supporting this.

However, note that the RISC-V specification explicitly allows implementations like this, where an SC never succeeds if the same hart executes a store in the meantime. RISC-V differentiates between _constrained_ and _unconstrained LR/SC loops_, where only the _constrained_ ones are guaranteed to succeed. From Section 8.3 in the RISC-V specification:

> The standard A extension defines constrained LR/SC loops, which have the following properties:
> - ...
> - An LR/SC sequence begins with an LR instruction and ends with an SC instruction. The
dynamic code executed between the LR and SC instructions can only contain instructions
from the base “I” instruction set, **excluding loads, stores**, backward jumps, taken backward
branches, JALR, FENCE, FENCE.I, and SYSTEM instructions. If the “C” extension is
supported, then compressed forms of the aforementioned “I” instructions are also permitted.
> - ...
>
> LR/SC sequences that do not lie within constrained LR/SC loops are unconstrained. Unconstrained
LR/SC sequences might succeed on some attempts on some implementations, **but might never
succeed on other implementations**.

So, while we can strengthen our implementation to allow more unconstrained LR/SC pairs to succeed, you still have to be careful with using stores in LR/SC loops. The stores can lead to livelocks, especially if there are regular writes to atomic variables/the reservation set. Also from Section 8.3:

> As a consequence of the eventuality guarantee, if some harts in an execution environment are
executing constrained LR/SC loops, and no other harts or devices in the execution environment
execute an unconditional store or AMO to that reservation set, then at least one hart will
eventually exit its constrained LR/SC loop. **By contrast, if other harts or devices continue to
write to that reservation set, it is not guaranteed that any hart will exit its LR/SC loop.**

> Software is not forbidden from using unconstrained LR/SC sequences, but portable software
must detect the case that the sequence repeatedly fails, then fall back to an alternate code sequence
that does not rely on an unconstrained LR/SC sequence. Implementations are permitted to
unconditionally fail any unconstrained LR/SC sequence.

**TLDR: While RISC-V allows an SC to succeed after a store from the same hart, it does not guarantee it, but even allows an implementation to always fail such an SC.**

**Comment by maxbjurling** [comment]
Thanks for the explanation! I understand now that the implementation is legal according to the RISC-V spec.
